### PR TITLE
feat: Add new Kakao Provider

### PR DESCRIPTION
**Pull request by esinx**
## What kind of change does this PR introduce?

This PR adds Kakao(https://accounts.kakao.com/) as an external provider.

## What is the current behavior?

This provider did not exist before.

## What is the new behavior?

Based on Kakao developer docs(https://developers.kakao.com/), this PR creates a provider & test suite for Kakao external provider.

## Additional context

Please let me know if there are any changes needed, I do acknowledge that this was once mentioned in another [comment](https://github.com/supabase/gotrue/issues/451#issuecomment-1101928384), but it seemed like the PR had been frozen since then. I wrote my own version to make sure the tests do pass and the features work properly.

### CODE REVIEW

File: api/provider/kakao.go
Comment (kangmingtay) [inline api/provider/kakao.go:80]: can you please provide a link to the available scopes that kakao support please?

File: api/provider/kakao.go
Comment (kangmingtay) [inline api/provider/kakao.go:49]: is it guaranteed that every kakao user will have an email? will the email still be returned even if it's not verified? why can't we just use `u.Account.EmailVerified` alone?

File: api/provider/kakao.go
Comment (kangmingtay) [inline api/provider/kakao.go:63]: what will this return if the user doesn't have a profile image url?

**Review by kangmingtay — CHANGES_REQUESTED** [review]
Hi @esinx, thanks for the PR, it would be great if you can provide a recording of some sort to show that this works (preferably via loom) and outline the steps needed to set this up. For example, it isn't immediately obvious where i should obtain the `GOTRUE_EXTERNAL_KAKAO_CLIENT_ID` and `GOTRUE_EXTERNAL_KAKAO_SECRET` from on the kakao developer portal.

File: api/provider/kakao.go
Comment (artifacts3) [reply api/provider/kakao.go:63]: null

File: api/provider/kakao.go
Comment (artifacts3) [reply api/provider/kakao.go:49]: https://developers.kakao.com/docs/latest/en/kakaologin/common

**Review by artifacts3 — APPROVED** [review]

File: api/provider/kakao.go
Comment (esinx) [reply api/provider/kakao.go:80]: Here's a screenshot of the list of scopes a project can request according to the Kakao developers documentation. Would simply adding the list of scopes in `oauthScopes` suffice?

Screen Shot 2023-01-06 at 11 21 24 PM

**Comment by esinx** [comment]
> 

Will post a loom within the next few days :) Thank you so much for reviewing this PR!

**Comment by esinx** [comment]
Here's a loom of me retrieving a new client ID & secret from the kakao developers page!
@kangmingtay @artifacts3 
https://www.loom.com/share/11bc4df13df848d7b4730f5e3906ca48

**Comment by igmyung** [comment]
May I know when the kakao login feature will be available for supabase auth? I need Kakao Login for a project in progress, and I want to refer to it for my release plan.

File: api/provider/kakao.go
Comment (kangmingtay) [reply api/provider/kakao.go:80]: hmm it seems like you need to minimally include the `account_email` and `profile_image` scope in order for kakao to return the user's email and profile image?

Comment (esinx) [reply api/provider/kakao.go]: That makes sense. I've addressed this issue on [8798d6f](https://github.com/supabase/gotrue/pull/834/commits/8798d6fba3e61db9af83b81232364dd1a9377d7d).

**Comment by esinx** [comment]
@kangmingtay I think this PR is ready to be reviewed-- is there an update on the feedbacks you left? I just pushed a sync that applies the project organization change(folder structure change).

@igmyung 
Meanwhile, you can use [this docker image](https://hub.docker.com/r/esinx/gotrue) if you intend to use Kakao as one of your providers. The tag you are looking for is `esinx/gotrue:kakao`.

**Review by kangmingtay — APPROVED** [review]

**Comment by kangmingtay** [comment]
hi @esinx, i think this is good to go! can you please merge in the master branch to fix the failing staticcheck and we can merge it in

**Comment by kangmingtay** [comment]
oh right, can you also do up some docs on how developers can use kakao as an oauth integration here (https://supabase.com/docs/guides/auth/overview#providers)? should be good as long as it's in the similar format as the rest! thanks @esinx !

**Comment by esinx** [comment]
Awesome! I'll do a short documentation write up asap.

**Review by artifacts3 — APPROVED** [review]

**Comment by kangmingtay** [comment]
hi @esinx, just circling back on this, have you had time to do up some docs for setting up oauth through kakao?

**Comment by J0** [comment]
Hey @esinx, 

Let us know if you need any help with pushing this through docs-wise - feel free to react ou via Discord (J0#7392) if there's any way we can be helpful

Jo

**Comment by esinx** [comment]
Hey! Sorry for being inactive-- I've been pretty busy with other work lately. I'll post an update on docs and also ask for some help on Discord for guidance. Thank you so much for all the support!

**Comment by hf** [comment]
@kangmingtay Could you try to review this one final pass and merge it?

**Comment by kangmingtay** [comment]
yup @hf we're waiting on the docs PR to be up before we merge it

**Comment by hwr12** [comment]
Hi @kangmingtay 
I wonder when will this PR be merged. 
And after it merges, when would I get to use this feature on supabase?

**Comment by kangmingtay** [comment]
Hi @hwr12, it will be merged once @esinx finishes the docs PR - feel free to help with this as well!

**Comment by kangmingtay** [comment]
Hi @hwr12, @esinx seems busy at the moment so feel free to pick up this task if you're interested. You can create the guide in this section of the supabase repo: https://github.com/supabase/supabase/tree/master/apps/docs/pages/guides/auth/social-login. Once you're done, please tag me in the pull request so that i can review it.

File: internal/conf/configuration.go
Comment (kangmingtay) [inline internal/conf/configuration.go:148]:
```suggestion
       Kakao                 OAuthProviderConfiguration `json:"kakao"`
```

File: internal/conf/configuration.go
Comment (kangmingtay) [inline internal/conf/configuration.go:148]:
```suggestion
	Google                  OAuthProviderConfiguration `json:"google"`
         Kakao                 OAuthProviderConfiguration `json:"kakao"`
```

File: internal/conf/configuration.go
Comment (kangmingtay) [inline internal/conf/configuration.go:148]:
```suggestion
	Google                  OAuthProviderConfiguration `json:"google"`
        Kakao                    OAuthProviderConfiguration `json:"kakao"`
```

**Comment by esinx** [comment]
My apologies for the delay. I just came back from my mandatory military service, and didn't have much time to go over the changes and the documentations.

@hwr12, thank you for taking the time to complete the documentation! This PR took longer than I've expected due to my unavailability, but I really do appreciate the community effort in trying to complete this PR.

Thank you @kangmingtay also for guiding us on how this PR should be made. Can't wait to see this feature make its way into supabase. Would be really amazing to have Kakao login available for all the Korean users out there.

**Comment by hwr12** [comment]
@kangmingtay 
Since the Kakao Auth from mobile supports Kakao Talk App sign in, it would be nice to add support for sign in with id token with Kakao in gotrue.
Do you guys have plans for this feature or will it be accepted when someone PRs this feature?

**Comment by kangmingtay** [comment]
Hi @hwr12, unfortunately not for now, the sign in with id token endpoint is still experimental but we can consider adding it in the future. We may also consider a webhook-based approach in the future which makes it easier to contribute oauth integrations to gotrue.

**Comment by hwr12** [comment]
@kangmingtay 
I really need this feature right now to continue using supabase as backend for my startup.
I can't think of any walk arounds of this feature lacking id token sign in method.
If PR dedicated to this feature is made, will you accept it? 
If you won't, would you mind telling me how I can achieve the same feature using other methods(eg. edge functions)?

**Comment by kangmingtay** [comment]
@hwr12 Can you explain why you are unable to use the `signInWithOAuth()` method instead?

**Comment by hwr12** [comment]
@kangmingtay Becuase calling `singInWithOAuth()` in flutter would open up browser and makes user to manually enter username and password which can be considered bothersome compared to one click app sign in.
To reduce steps of sign in process, the app should directly interact with Kakao Talk app and retrieve identity token, and using the token to sign in or sign up to my service.

**Comment by esinx** [comment]
I've achieved such behavior through modifying the link generated by GoTrue. See https://github.com/esinx/RN-Kakao-Login-Gotrue

**Comment by hwr12** [comment]
Hi @esinx !
Are you referring to this code in the below picture?
Google Chrome-2023 05 19 16 49-001178
Could you explain in detail if you don't mind?

**Comment by hwr12** [comment]
@esinx I am confused due to my lack of knowledge but after the login is handled on the Kakao Talk app, is the below code used to continue on signing in to supabase server?

Google Chrome-2023 05 19 17 07-001180

This code looks like doing it manually as I see refresh token and access token are used to set session. 
If I understood correctly, the Kakao talk login handling emits refresh token and access token for supabase authentication that are used to set session manually in the app?

**Comment by hf** [comment]
Hey everyone looking for the ID Token flow. First we need to finish #1108 and then we can add support for Kakao ID tokens.

**Comment by hwr12** [comment]
Thank you @hf .
I really need this feature ASAP.
Hope to see this feature coming to supabase soon!!

**Comment by z3rosmith** [comment]
@hf @kangmingtay Is there any update on the support for Kakao ID tokens?

**Pull request merged**
